### ISSUE TITLE
[OJS] Setting delayed open access for new issue randomly fails (settings file cache problem?)

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2888]
Hi @asmecher,

Yes, I know that the embargo settings are not available until 3.0.2. But basically you can use them by giving the correct settings in journal_settings table (publishingMode = 1; enableDelayedOpenAccess = 1 and delayedOpenAccessDuration = 12). 

With these three settings set, we have been succesful in setting embargos for new issues. The related code is here: https://github.com/pkp/ojs/blob/ojs-stable-3_0_2/classes/controllers/grid/issues/IssueGridHandler.inc.php#L441-L456. 

However, today one of our journals using the embargo settings published a new issue, but the embargo was not set for the issue and it's articles. The access status was not set and the open access date was null. The code is almost the same in OJS 3.1 so this could be a problem with the future version as well.

I, among other users, received a notification for the new issue, meaning that the code has reached here: https://github.com/pkp/ojs/blob/ojs-stable-3_0_2/classes/controllers/grid/issues/IssueGridHandler.inc.php#L463

But for some reason, the conditions given here were not met: https://github.com/pkp/ojs/blob/ojs-stable-3_0_2/classes/controllers/grid/issues/IssueGridHandler.inc.php#L441 

The same journal has published a new issue two times before using OJS3.0.2 and both times the embargo was set succesfully. We have not changed their settings and we have not changed the OJS code.

The only explanation I can think of is connected to these two errors I found that seem to match the time of publishing:
```
[Wed Oct 18 11:39:03 2017] [error] [client 128.214.149.46] PHP Warning: include(/var/www/http://www.site.fi/public_html/cache/fc-pluginSettings-29-introduction.php): failed to open stream: File or folder does not exist in /var/www/http://www.site.fi/public_html/lib/pkp/classes/cache/FileCache.inc.php on line 46, referer: https://site.fi/journalname/manageIssues
[Wed Oct 18 11:39:03 2017] [error] [client 128.214.149.46] PHP Warning: include(): Failed opening '/var/www/http://www.site.fi/public_html/cache/fc-pluginSettings-29-introduction.php' for inclusion (include_path='.:/var/www/http://www.site.fi/public_html/classes:/var/www/http://www.site.fi/public_html/pages:/var/www/http://www.site.fi/public_html/lib/pkp:/var/www/http://www.site.fi/public_html/lib/pkp/classes:/var/www/http://www.site.fi/public_html/lib/pkp/pages:/var/www/http://www.site.fi/public_html/lib/pkp/lib/adodb:/var/www/http://www.site.fi/public_html/lib/pkp/lib/phputf8:/var/www/http://www.site.fi/public_html/lib/pkp/lib/pqp/classes:/var/www/http://www.site.fi/public_html/lib/pkp/lib/smarty:.:/usr/share/pear:/usr/share/php') in /var/www/http://www.site.fi/public_html/lib/pkp/classes/cache/FileCache.inc.php on line 46, referer: https://site.fi/journalname/manageIssues
```
The `fc-pluginSettings-29-introduction.php` is the cache file for a simple custom block plugin. 

Not related as such, but **is it possible that the encoutered error has affected somehow the way OJS has read the journal settings from the file cache**? I mean, is it possible that OJS could not read the journal settings because of that error? I know, a far reach, but really the only thing I could think of.

**Comment by asmecher** [comment]
Hmm, it's hard to diagnose this after the fact and at a distance, but one common cause of trouble when working with database entries for `journal_settings` manually: if the setting is changed, but the file cache is not flushed, then you'll have a time bomb: the settings will behave per the cache, but as soon as *any* journal setting changes, or the cache file is manually deleted, the whole cache will be re-generated from the database and the new settings will take effect.

Some of the settings involved in that code, e.g. `publishingMode`, will be especially volatile because there's no way to save the settings form without bumping it out of the mode you'd specified via a manual DB intervention.

**Comment by ajnyga** [comment]
Yes, I have had such problems with the cache, but I do not think it was the reason in this case.

- I changed those settings back in January and the journal has published twice successfully after that.
- I also checked the cache file right after this happened and the settings are correct there as well
- Also with the publishingmode setting, I just took back this change you made in last August so the journal has had a working form since January: https://github.com/pkp/ojs/commit/1baa43534198aa87081b54bfcc94c38954c36670#diff-8328578bce24fcf41aeedf58bb476dea

**Comment by asmecher** [comment]
@ajnyga, have you upgraded to OJS 3.1? If so, is it possible to watch for this to reappear on that codebase?

**Comment by ajnyga** [comment]
Hi @asmecher, no I have not yet.

I can not see any logical reason why the code setting the embargo has failed. Other than the possibility that the journal has changed their settings themselves. The code in 3.1 is almost the same.

But since the journals can contol the embargos themselves starting from 3.1 I can just ask them to always check the embargo settings for new issues once they are published. 

But I do think that the file cache errors are something that needs a closer look. We get those a lot. Most of them are related to block plugins but others exist too.

**Comment by asmecher** [comment]
@ajnyga, we get warnings over here (mostly about file existence). It appears to be a race condition in the file caching code. However, it seems to be cosmetic -- I've never seen it causing problems beyond cluttering the error log. Are you seeing something different?

By the way, our plans for the caching code are to *stop maintaining our own* :) and integrate something like [Stash](https://github.com/tedious/Stash). As long as the warnings are just cosmetic, I'm more interested in replacing the caching code than tweaking it.

**Comment by ajnyga** [comment]
@asmecher, I think that they are all about missing files.

If it is just a race condition thing and nothing critical happens, then it is fine by me. Although it would be nice to have an empty error log :-) 

Third party solution sounds nice, but it is probably not among your top priorities.

**Comment by ajnyga** [comment]
@asmecher I probably found a solution to this issue.

It is this old code here (not a problem anymore in 3.1): https://github.com/pkp/ojs/blob/ojs-stable-3_0_2/controllers/grid/issues/form/IssueForm.inc.php#L223-L225

Basically everytime, after publishing the issue, you save the issue form in OJS3.0.2 it will change the accessStatus to open (1) and remove the openAccesDate. I did a small fix while waiting for 3.1.1.

Anyway, closing this because I am quite sure the journal just edited something in the issue details after publishing it.

**Comment by asmecher** [comment]
Ah, thanks for following up, @ajnyga! That makes sense.
